docs/contributors/coding-guidelines.md

```diff
@@ -195,6 +195,10 @@ alert( 'My name is ' + name + '.' );
 alert( `My name is ${ name }.` );
 ```
 
+### `@wordpress/element` (React) Components
+
+It is preferred to implement all components as [function components](https://reactjs.org/docs/components-and-props.html), using [hooks](https://reactjs.org/docs/hooks-reference.html) to manage component state and lifecycle. It's not expected that you should ever encounter a situation where you must use a class component. Note that the [WordPress guidance on Code Refactoring](https://make.wordpress.org/core/handbook/contribute/code-refactoring/) applies here: There needn't be a concentrated effort to update class components in bulk. Instead, consider it as a good refactoring opportunity in combination with some other change.
```

Comment:
> It's not expected that you should ever encounter a situation where you must use a class component.

Actually, there is. Error boundaries.

Reply:
> Actually, there is. Error boundaries.

Good call. It's probably a very rare occurrence for most developers they'd be using error boundaries. All the same, I guess I could just remove this sentence to avoid the inaccuracy, and to avoid going into all of the specifics (the exception)?

For context, my rationale with including a sentence like this was to try account for scenarios where a developer feels that hooks "aren't enough" to recreate a behavior they're otherwise familiar with in a class implementation, in order to create resistance where they may otherwise feel inclined that a class component must be used.

Reply:
Alternatively, we can just mention the exception. I'm fine either way though.

Reply:
Perhaps:

> With a few specific exceptions (such as error boundaries), you will almost never encounter a situation where you must use a class component.

Reply:
@chrisvanpatten I was thinking something like this. Looking around, I believe error boundaries really are the _only_ exception here ([unofficial reference](https://gist.github.com/andywer/800f3f25ce3698e8f8b5f1e79fed5c9c)).

So it could be as simple as:

"With the exception of error boundaries, ..."

Reply:
Added mention of the exception in fedc81b.